### PR TITLE
feature: Sorting & finding by username (ID) admin interface  

### PR DESCRIPTION
**Pull request by owen-duncan-snobel**
Feature changes for https://github.com/LAION-AI/Open-Assistant/issues/1615 

- Auto-focuses on filter tab
- Able to sort / filter by username (ID)

https://user-images.githubusercontent.com/43126781/220240282-5ed28ca1-1514-4008-addd-96ea6b8a443d.mov

### CODE REVIEW

File: website/src/components/DataTable/DataTable.tsx [hunk 1]
```diff
@@ -35,7 +35,7 @@ import {
 } from "@tanstack/react-table";
 import { Filter } from "lucide-react";
 import { useTranslation } from "next-i18next";
-import { ChangeEvent, ReactNode, useState } from "react";
+import { ChangeEvent, Dispatch, ReactNode, SetStateAction, useRef, useState } from "react";
```
Comment (AbdBarho) [inline website/src/components/DataTable/DataTable.tsx:38]: are all of these imports used?

File: website/src/components/DataTable/DataTable.tsx [hunk 2]
```diff
@@ -198,13 +200,13 @@ const FilterModal = ({
   value: string;
 }) => {
   const { isOpen, onOpen, onClose } = useDisclosure();
-
+  const initialFocusRef = useRef();
```
Comment (AbdBarho) [inline website/src/components/DataTable/DataTable.tsx:203]: what is the purpose of this ref?

**Review by AbdBarho — COMMENTED** [review]
Thank you

File: website/src/components/DataTable/DataTable.tsx (hunk 1 above)
Comment (owen-duncan-snobel) [reply website/src/components/DataTable/DataTable.tsx:38]: @AbdBarho Not all of them are used, I'll remove them now. My first attempt used Dispatch and SetStateAction but I changed approaches at some point

File: website/src/components/DataTable/DataTable.tsx (hunk 2 above)
Comment (owen-duncan-snobel) [reply website/src/components/DataTable/DataTable.tsx:203]: @AbdBarho That ref is for auto-focusing on the text input when the filter button is clicked. That way you can start typing as soon as the parent div/button is clicked

File: website/src/components/DataTable/DataTable.tsx [hunk 3]
```diff
@@ -59,7 +61,7 @@ export type DataTableProps<T> = {
   filterValues?: FilterItem[];
   onNextClick?: () => void;
   onPreviousClick?: () => void;
-  onFilterChange?: (items: FilterItem[]) => void;
+  onFilterChange?: (items: FilterItem[], filterByColumn?: string) => void;
```
Comment (notmd) [inline website/src/components/DataTable/DataTable.tsx:62]: Why do you need filterByColumn here? Isn't `id` in `FilterItem` enough?

File: website/src/components/DataTable/DataTable.tsx [hunk 4]
```diff
@@ -50,6 +50,8 @@ export type FilterItem = {
   value: string;
 };
 
+export type FilterOption = "display_name" | "username";
```
Comment (notmd) [inline website/src/components/DataTable/DataTable.tsx]: I'm not sure why we need this, but it seem incorrect. The DataTable component is generic, it should not know column filter name

File: website/src/components/UserTable.tsx [hunk 5]
```diff
@@ -47,17 +50,23 @@ const columns: DataTableColumnDef<User>[] = [
 export const UserTable = memo(function UserTable() {
   const { pagination, resetCursor, toNextPage, toPreviousPage } = useCursorPagination();
   const [filterValues, setFilterValues] = useState<FilterItem[]>([]);
-  const handleFilterValuesChange = (values: FilterItem[]) => {
+  const [filterOption, setFilterOption] = useState<string>("display_name");
```
Comment (notmd) [inline website/src/components/UserTable.tsx:53]: You dont need this state. The last FilterItem is the column you want, so just discard the rest values and keep the last one only

File: website/src/components/DataTable/DataTable.tsx (hunk 4 above)
Comment (owen-duncan-snobel) [reply website/src/components/DataTable/DataTable.tsx]: Ya sorry, I have removed that type

File: website/src/components/UserTable.tsx (hunk 5 above)
Comment (owen-duncan-snobel) [reply website/src/components/UserTable.tsx]: I went and removed the state and used the last FilterItem

File: website/src/components/DataTable/DataTable.tsx (hunk 3 above)
Comment (owen-duncan-snobel) [reply website/src/components/DataTable/DataTable.tsx:62]: I went and removed it

File: website/src/components/UserTable.tsx [hunk 6]
```diff
@@ -55,9 +58,18 @@ export const UserTable = memo(function UserTable() {
   // Fetch and save the users.
   // This follows useSWR's recommendation for simple pagination:
   //   https://swr.vercel.app/docs/pagination#when-to-use-useswr
-  const display_name = filterValues.find((value) => value.id === "display_name")?.value ?? "";
+
+  // TODO this should probably change names + back-end changes since we are search by not only display name
```
Comment (notmd) [inline website/src/components/UserTable.tsx:62]: Can you rename the query to something like filterValue and remove this todo?

File: website/src/components/DataTable/DataTable.tsx [hunk 7]
```diff
@@ -59,7 +59,7 @@ export type DataTableProps<T> = {
   filterValues?: FilterItem[];
   onNextClick?: () => void;
   onPreviousClick?: () => void;
-  onFilterChange?: (items: FilterItem[]) => void;
+  onFilterChange?: (items: FilterItem[], filterByColumn?: string) => void;
```
Comment (notmd) [inline website/src/components/DataTable/DataTable.tsx:62]:
```suggestion
  onFilterChange?: (items: FilterItem[]) => void;
```

**Review by notmd — CHANGES_REQUESTED** [review]
Thanks, look good now

**Comment by notmd** [comment]
Ah you need to update the `handleFilterValuesChange`, some thing like might work (please test it)
```
const handleFilterValuesChange = (values: FilterItem[]) => {
  const last = values.pop()
  if (last) {
    setFilterValues([last])
  }
  resetCursor();
}
```

**Review by notmd — APPROVED** [review]
Thanks

**Pull request merged**